### ISSUE TITLE
reading large files fails

### ISSUE THREAD

**Issue by r5r3** [irods/python-irodsclient#280]
The following code opens a file and reads the complete content. That works well for smaller files (< 1GB), but fails with larger files:

```
obj = session.data_objects.get(remote_path)
with open(local_filepath, "wb") as local_file, obj.open() as remote_file:
    position = 0
    while position < obj.size:
        read_bytes = min(session.data_objects.READ_BUFFER_SIZE, obj.size - position)
        file_chunk = remote_file.read(read_bytes)
        bytes_read = len(file_chunk)
        local_file.write(file_chunk)
        position += bytes_read
```

At some point, the `read` method will begin to return empty results. 

Using `session.data_objects.get(remote_path, local_path)` to download a large file fails as well (with `parallel get failed`). With `num_threads=1` `get` will not fail, by the file will also not be downloaded completely. 

Tested with python-irodsclient version 1.0.0 and 0.9.0 and iRODS Server version 4.2.9. The problem seems to be related to the iRODS server version. Downloading large files with python and the server version 4.2.8 used to work perfectly.

**Comment by r5r3** [comment]
Additional information: downloading large files using `iget` is not a problem.

**Comment by trel** [comment]
Thanks for the report, we'll reproduce and box it in.

**Comment by d-w-moore** [comment]
@r5r3 Could you send me more info about your setup?  I presume you're running release v1.0.0 of the python-irodsclient.  
Have you tried pointing your client at a 4.2.8 server to see if the  issue persists? 
Are there error messages produced on the console and/or in the iRODS server log?

 I tried your  above script,  and it worked fine for me,  simply downloading a 100M data object (on a unixfilesystem resource) into a local file.

Is possible that the replica you are trying to download is on an S3 resource,or something else other than a unix or posix filesystem?

**Comment by d-w-moore** [comment]
Ok I've reproduced it.  Evidently its an 0x7fffffff (32 bit signed ) problem....

**Comment by d-w-moore** [comment]
This appears to be an issue in iRODS 4.2.9, and this would seem to solve it: https://github.com/irods/irods/pull/5708
Will submit through review and testing, and the fix (or something like it ; ) ) should be out in 4.2.10.
I am not sure, however, if there will be a workaround that allows iRODS clients to download files in the size range of 2G - 4G, 6G-8G,  in the meantime.
```
$ dd if=/devu/random count=2100k bs=1k of=junk.put
$ iput -f junk.put
$ python -c "from irods.test.helpers import make_session
session = make_session()
remote_path = '/tempZone/home/rods/junk.put'
local_filepath = 'junk.get'
obj = session.data_objects.get(remote_path)
position = 0
with open(local_filepath, "wb") as local_file, obj.open() as remote_file:
    while position < obj.size:
        read_bytes = min( obj.size - position, session.data_objects.READ_BUFFER_SIZE )
        file_chunk = remote_file.read(read_bytes)
        bytes_read = len(file_chunk)
        local_file.write(file_chunk)
        position += bytes_read
        print(bytes_read)
" | tail -5
$ diff junk.{get,put} && echo Yes
Yes
```

**Comment by d-w-moore** [comment]
As one might expect, this does not just apply to the python client. 
Note the comments to see what number-of-bytes-read-out values are produced in the serverLog ...
```                                             
read_test() {                                                                          
       msiDataObjOpen("objPath=/tempZone/home/rods/junk.put", *fd);   # a data object of 2100 MB (2.1G)
*len = 2048;    
        msiDataObjRead(*fd, *len, *buf); 
        writeLine("serverLog", "--- size of data block read  ---   ");                              
        writeLine("serverLog", *buf);  
                         # -> "0" on 4.2.9 for a data object of size [ > 2G, < 4G ] but "2048" for other objects >= 2048 bytes    
                         # -> "2048" on 4.2.8 for any known size of data object       >= 2048 bytes                            
        msiDataObjClose(*fd, *close_out);
}
```

**Comment by r5r3** [comment]
Thanks a lot for finding the source of the issue!

> I am not sure, however, if there will be a workaround that allows iRODS clients to download files in the size range of 2G - 4G, or 6G - 8G, etc., in the meantime.

For me, I is only a problem for the python client. With `iget`, I can download files in the range of 2 to 4 GB without any data corruption.

**Comment by trel** [comment]
This appears due to the new check provided to fix https://github.com/irods/irods/issues/5352 in 4.2.9.

**Comment by trel** [comment]
This will be marked as duplicate of https://github.com/irods/irods/issues/5709 in the server itself.

**Comment by trel** [comment]
Dan, can you confirm that this is fixed against irods/irods 4-2-stable?  If so, please close.

**Comment by d-w-moore** [comment]
I'm sure but will verify and then close.

**Comment by d-w-moore** [comment]
Fixed in irods/irods#5709
